memq-client/src/main/java/com/pinterest/memq/client/commons2/MemqCommonClient.java

```diff
@@ -197,12 +288,115 @@ protected List<Endpoint> randomizedEndpoints(List<Endpoint> servers) {
+   * to the set of writeEndpoints until the size of writeEndpoints reaches numWriteEndpoints. An example is provided in javadoc for sendRequestPacketAndReturnResponseFuture().
+   * 
+   * @return the endpoints to try
+   */
+  protected List<Endpoint> getEndpointsToTry() {
+    List<Endpoint> endpointsToTry = new ArrayList<>();
+
+    if (writeEndpoints.size() == numWriteEndpoints) {
+      // write endpoints is full, rotate it by 1 and concatenate with the remaining locality endpoints
+      Collections.rotate(writeEndpoints, 1);
+      endpointsToTry.addAll(new ArrayList<>(writeEndpoints));
+      List<Endpoint> remainingEndpoints = new ArrayList<>(localityEndpoints);
+      remainingEndpoints.removeAll(writeEndpoints);
+      endpointsToTry.addAll(remainingEndpoints);
+    } else {
+      // write endpoints is not full, rotate the locality endpoints by 1
+      Collections.rotate(localityEndpoints, -1);
+      endpointsToTry.addAll(localityEndpoints);
+    }
+
+    return endpointsToTry;
+  }
+
+  /**
+   * The provided endpoint had just succeeded, so register the endpoint as a write endpoint if it is not already in the set of write endpoints and if the set of write endpoints is not full.
+   * 
+   * The endpoint's failure count is reset to 0 since it had just succeeded.
+   * 
+   * If the set of write endpoints is full, nothing is done.
+   * 
+   * @param endpoint
+   * @param topic
+   */
+  protected void maybeRegisterWriteEndpoint(Endpoint endpoint, String topic) {
+    failureCounts.remove(endpoint);
+    if (writeEndpoints.size() < numWriteEndpoints && !writeEndpoints.contains(endpoint)) {
+      logger.info("Registering write endpoint: " + endpoint + " for topic: " + topic);
+      writeEndpoints.add(endpoint);
+    }
+    if (!localityEndpoints.contains(endpoint)) {
+      // we should not reach here, because the endpoint should be in the set of locality endpoints. Adding this logic as a sanity check / fail-safe
+      logger.info("Registering locality endpoint: " + endpoint + " for topic: " + topic);
+      localityEndpoints.add(endpoint);
+    }
+  }
+
+  /**
+   * Deprioritize the dead endpoint by removing it from the set of write endpoints and moving it to the end of locality endpoints.
+   * 
+   * If the endpoint has failed 2 times in a row, it is removed from both the set of write endpoints and the set of locality endpoints so it is not considered for future requests.
+   * 
+   * If the endpoint has failed 1 time but succeeds again in a future request, its failure count is reset to 0 in maybeRegisterWriteEndpoint().
+   * 
+   * An example is provided in javadoc for sendRequestPacketAndReturnResponseFuture().
+   * 
+   * @param deadEndpoint
+   * @param topic
+   * @throws Exception
+   */
+  protected void deprioritizeDeadEndpoint(Endpoint deadEndpoint, String topic) throws Exception {
+      // put deadEndpoint last in the priority
+      failureCounts.compute(deadEndpoint, (k, v) -> v == null ? 1 : v + 1);
```

Comment:
Does `failureCounts` ever decrease? In case, bad brokers become healthy again?

Reply:
the failure count is reset to 0 (i.e. removed from the map) in `maybeRegisterWriteEndpoint` which is called if the endpoint was successful: https://github.com/pinterest/memq/blob/9cb7ce905a045a7a306c9693ae0b60db8922e55b/memq-client/src/main/java/com/pinterest/memq/client/commons2/MemqCommonClient.java#L351